### PR TITLE
Require Java 8 and include some Java 8 dependencies.

### PR DESCRIPTION
**Pull request by gianm**
Patch for #3746, #3745, #3710, #3595.

- Upgrade Jetty to 9.3.16.v20170120.
- Upgrade DataSketches to 0.8.4.
- Bundle caffeine-cache by default.
- Still target Java 7 when compiling base Druid classes.

Our hadoop gauntlet tests don't pass with this, because they are running on Java 7. But hopefully they will pass after upgrading that test environment to Java 8. In the meantime I'm doing this as a WIP PR just so people can see the changes and try it in their own environment if they want.

### CODE REVIEW

**Comment by gianm** [comment]
Shall we also disable travis builds for jdk7? They, of course, do not pass after applying this patch.

**Comment by jon-wei** [comment]
I tested this PR for hadoop compatibility with:
* sequenceiq hadoop docker image 2.3.0 to 2.7.3
* cloudera quickstart docker image latest
* hortonworks hdp 2.5 docker image
* an EMR 5.0.0 cluster

all running java8 on the hadoop cluster side, running a batch ingestion task on the wikiticker data, those all passed

**Comment by gianm** [comment]
Thanks @jon-wei! In that case I'll take off the WIP tag and this is ready for review.

File: extensions-core/datasketches/pom.xml [hunk 1]
```diff
@@ -38,7 +38,7 @@
     <dependency>
       <groupId>com.yahoo.datasketches</groupId>
       <artifactId>sketches-core</artifactId>
-      <version>0.4.1</version>
+      <version>0.8.4</version>
```
Comment (drcrallen) [inline extensions-core/datasketches/pom.xml]: Is this really germane to java 8?

**Comment by drcrallen** [comment]
Discussed in dev sync. @gianm was going to update travis yaml before merge

**Comment by gianm** [comment]
Removed jdk7 from travis.yml.

**Comment by fjy** [comment]
👍

**Pull request merged**